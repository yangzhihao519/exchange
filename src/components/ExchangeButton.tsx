import React from 'react'

interface ExchangeButtonProps {
    onClickExchange: () => void,
    disabled: boolean
}

class ExchangeButton extends React.Component<ExchangeButtonProps>{

    render() {
        let className = this.props.disabled? 'exchange-button primary-button disabled' : 'exchange-button primary-button'

        return (<div className={className} onClick={this.props.onClickExchange}> 
            Exchange
        </div>)
    }
    
}

export default ExchangeButton