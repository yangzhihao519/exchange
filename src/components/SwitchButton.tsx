import React from 'react'

interface SwitchButtonProps {
    onSwitchExchange: () => void
}

class SwitchButton extends React.Component<SwitchButtonProps>{

    render() {
        return (<div className='switch-button' onClick={this.props.onSwitchExchange} />)
    }
}

export default SwitchButton