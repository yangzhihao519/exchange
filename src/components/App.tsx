import React from 'react';
import '../css/styles.css';
import Area from './Area';
import Money from '../models/money'
import Currency from '../models/currency';
import ExchangeButton from '../components/ExchangeButton'
import SwitchButton from '../components/SwitchButton'
import ExchangeRate from '../components/ExchangeRate'

interface AppProps {
  initBalances: Money[]
  exchange: {
    sellOut: Money,
    buyIn: Money
  }
  allCurrencies: Currency[]
  onInputChange: (value: number, isSellingOut: boolean) => void
  onClickExchange: () => void
  onSwitchExchange: () => void
  onSelectCurrency: (selectedCurrencyCode: string, isSellingOut: boolean) => void
  fetchExchangeRate: () => void
}

interface AppState {}

class App extends React.Component<AppProps, AppState>{
  
  timer = 0

  componentDidMount() {
    this.props.fetchExchangeRate()
    // Send the API every 10 seconds to refresh the rates
    // The line below was commented out due to the limitation of API calls per month
    //this.timer = setInterval(() => this.props.fetchExchangeRate(), 10000);
  }

  componentWillUnmount() {
      clearInterval(this.timer);
      this.timer = 0;
  }

  render() {
    let { initBalances, exchange, onClickExchange, onSwitchExchange, allCurrencies} = this.props
    let sellOutExchange = exchange.sellOut
    let sellOutInitBalance = initBalances.find((item) => {return item.currencyCode === sellOutExchange.currencyCode})
    let buyInExchange = exchange.buyIn
    let buyInInitBalance = initBalances.find((item) => {return item.currencyCode === buyInExchange.currencyCode})
    let sellOutCurrency = allCurrencies.find((item) => {return item.code === sellOutExchange.currencyCode})
    let buyInCurrency = allCurrencies.find((item) => {return item.code === buyInExchange.currencyCode})
    let disableExchangeButton = sellOutInitBalance ? sellOutInitBalance.value < sellOutExchange.value || sellOutExchange.value === 0 || sellOutExchange.currencyCode === buyInExchange.currencyCode: true

    return(
      <div className='app'>
        <div className='app__title'>Exchange</div>
        <div className='area-container'>
            {sellOutInitBalance ? <Area allCurrencies={allCurrencies}
                                        initBalance={sellOutInitBalance} 
                                        exchange={sellOutExchange} 
                                        onInputChange={this.props.onInputChange} 
                                        onSelectCurrency={this.props.onSelectCurrency}
                                        isSellingOut={true}/> 
                                :  <div>Sell Out Balance is not available.</div>}
            <div className='area__middle'>
              <SwitchButton onSwitchExchange={onSwitchExchange}/>
              {sellOutCurrency && buyInCurrency && <ExchangeRate sellOutCurrency={sellOutCurrency} buyInCurrency={buyInCurrency}/>}
            </div>
            {buyInInitBalance ? <Area allCurrencies={allCurrencies}
                                      initBalance={buyInInitBalance} 
                                      exchange={buyInExchange} 
                                      onInputChange={this.props.onInputChange} 
                                      onSelectCurrency={this.props.onSelectCurrency}
                                      isSellingOut={false}/> 
                              : <div>Buy In Balance is not available.</div>}
          </div>
          <ExchangeButton onClickExchange={onClickExchange} disabled={disableExchangeButton}/>
      </div>
    )
  }
}

export default App;
