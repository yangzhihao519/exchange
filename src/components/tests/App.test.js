import React from 'react';
import App from '../App';
import { mount, shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import SwitchButton from '../SwitchButton';
import ExchangeButton from '../ExchangeButton';

const mockFunction = jest.fn();

describe('App', () => {
  it('renders without crashing given the required props', () => {
    const props = {
      initBalances: [],
      exchange: {},
      allCurrencies: [],
      onInputChange: mockFunction,
      onClickExchange: mockFunction,
      onSwitchExchange: mockFunction,
      onSelectCurrency: mockFunction,
      fetchExchangeRate: mockFunction
    }
    const wrapper = shallow(<App {...props} />)
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('should call mockFunction on Switch button click', () => {
    const component = mount(
      <SwitchButton onSwitchExchange={mockFunction} />
    );
    component.find('.switch-button').simulate('click');
    expect(mockFunction).toHaveBeenCalled();
    
    component.unmount();
  });
  
  it('should call mockFunction on Exchange button click', () => {
    const component = mount(
      <ExchangeButton onClickExchange={mockFunction} disabled={false} />
    );
    component.find('.exchange-button').simulate('click');
    expect(mockFunction).toHaveBeenCalled();
    
    component.unmount();
  });
})