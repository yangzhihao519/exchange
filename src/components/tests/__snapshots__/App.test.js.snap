// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders without crashing given the required props 1`] = `
<div
  className="app"
>
  <div
    className="app__title"
  >
    Exchange
  </div>
  <div
    className="area-container"
  >
    <div>
      Sell Out Balance is not available.
    </div>
    <div
      className="area__middle"
    >
      <SwitchButton
        onSwitchExchange={
          [MockFunction] {
            "calls": Array [
              Array [],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": undefined,
              },
            ],
          }
        }
      />
    </div>
    <div>
      Buy In Balance is not available.
    </div>
  </div>
  <ExchangeButton
    disabled={true}
    onClickExchange={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      }
    }
  />
</div>
`;
